core/oa_hash_map.h

```diff
@@ -0,0 +1,571 @@
+					// because we know it's not in here.
+
+					check_old_table = false;
+				}
+			}
+
+			if (check_new_table) {
+
+				int pos = (hash + i) % table.capacity;
+
+				int flags_pos = pos / 4;
+				int flags_pos_offset = pos % 4;
+
+				bool is_filled_flag = (table.flags[flags_pos] & (1 << (2 * flags_pos_offset))) > 0;
+				bool is_deleted_flag = (table.flags[flags_pos] & (1 << (2 * flags_pos_offset + 1))) > 0;
+
+				if (is_filled_flag) {
+					// found our entry?
+					if (table.hashes[pos] == hash && Comparator::compare(table.keys[pos], p_key)) {
+						table.keys[pos].~TKey();
+						table.data[pos].~TData();
+
+						memnew_placement(&table.keys[pos], TKey);
+						memnew_placement(&table.data[pos], TData);
+
+						table.flags[flags_pos] &= ~(1 << (2 * flags_pos_offset));
+						table.flags[flags_pos] |= (1 << (2 * flags_pos_offset + 1));
+
+						// don't return here, this value might still be in the old table
+						// if it was already relocated.
+
+						elements--;
+						return;
+					}
+					continue;
+				} else if (is_deleted_flag) {
+					continue;
+				} else {
+					check_new_table = false;
+				}
+			}
+		}
+	}
+
+	struct Iterator {
+		bool valid;
+
+		uint32_t hash;
+
+		const TKey *key;
+		const TData *data;
+
+	private:
+		friend class OAHashMap;
+		bool was_from_old_table;
+	};
+
+	Iterator iter() const {
+		Iterator it;
+
+		it.valid = false;
+		it.was_from_old_table = false;
+
+		bool check_old_table = is_rehashing;
+
+		for (int i = 0; i < table.capacity; i++) {
+
+			// if we're rehashing check the old table first
+			if (check_old_table && i < old_table.capacity) {
+
+				int pos = i;
+
+				int flags_pos = pos / 4;
+				int flags_pos_offset = pos % 4;
+
+				bool is_filled_flag = (old_table.flags[flags_pos] & (1 << (2 * flags_pos_offset))) > 0;
+
+				if (is_filled_flag) {
+					it.valid = true;
+					it.hash = old_table.hashes[pos];
+					it.data = &old_table.data[pos];
+					it.key = &old_table.keys[pos];
+
+					it.was_from_old_table = true;
+
+					return it;
+				}
+			}
+
+			{
+
+				int pos = i;
+
+				int flags_pos = pos / 4;
+				int flags_pos_offset = pos % 4;
+
+				bool is_filled_flag = (table.flags[flags_pos] & (1 << (2 * flags_pos_offset))) > 0;
+
+				if (is_filled_flag) {
+					it.valid = true;
+					it.hash = table.hashes[pos];
+					it.data = &table.data[pos];
+					it.key = &table.keys[pos];
+
+					return it;
+				}
+			}
+		}
+
+		return it;
+	}
+
+	Iterator next_iter(const Iterator &p_iter) const {
+		if (!p_iter.valid) {
+			return p_iter;
+		}
+
+		Iterator it;
+
+		it.valid = false;
+		it.was_from_old_table = false;
+
+		bool check_old_table = is_rehashing;
+
+		// we use this to skip the first check or not
+		bool was_from_old_table = p_iter.was_from_old_table;
+
+		int prev_index = (p_iter.data - (p_iter.was_from_old_table ? old_table.data : table.data));
+
+		if (!was_from_old_table) {
+			prev_index++;
+		}
+
+		for (int i = prev_index; i < table.capacity; i++) {
```

Comment:
++i maybe?

Reply:
No need to duplicate this comment :wink:
IMO i++ is fine, given how many times we used it throughout the code.

Reply:
Yeah, figured it'd be easier to find for @karroffel if I noted all occurrences, maybe not *that* helpful.